File: android/shared/src/commonMain/kotlin/com/chaekchaek/app/domain/book/Book.kt
package com.chaekchaek.app.domain.book

/**
 * 책 한 권. Figma 검색 결과와 책 상세에 쓰이는 서지 정보다.
 *
 * ```
 * 마션                                    title
 * 앤디 위어 · 박아람 옮김                  authors, translators
 * 알에이치코리아 · SF · 2026 · 308쪽       publisher, category, publishedYear, totalPages
 * ```
 *
 * 저자와 역자를 나눈 이유는 검색 결과가 "옮김"을 붙여 구분하고, 서재 목록은 저자만 쓰기
 * 때문이다. 표시 문자열 조합은 presentation 이 담당한다.
 */
class Book(
    val id: BookId,
    val title: String,
    val authors: List<String>,
    val translators: List<String> = emptyList(),
    val publisher: String,
    val category: String,
    val publishedYear: Int,
    val totalPages: PageCount,
    val coverId: String,
) {
    init {
        require(title.isNotBlank()) { "책 제목은 공백일 수 없습니다." }
        require(authors.isNotEmpty()) { "저자가 최소 한 명 필요합니다." }
    }
}

File: android/iosApp/Chaekchaek/Book.swift
import Foundation

enum ReadingStatus: String, Codable, CaseIterable, Identifiable, Sendable {
    case wantToRead
    case reading
    case finished

    var id: Self { self }

    var title: String {
        switch self {
        case .wantToRead: "읽고 싶어요"
        case .reading: "읽는 중"
        case .finished: "다 읽음"
        }
    }
}

struct Book: Codable, Hashable, Identifiable, Sendable {
    let isbn13: String
    let isbn: String
    let title: String
    let author: String
    let publisher: String
    let coverURL: URL?
    let category: String
    let publishedAt: String
    let totalPages: Int
    let description: String

    var id: String {
        if !isbn13.isEmpty { return isbn13 }
        if !isbn.isEmpty { return isbn }
        return "\(title)|\(author)|\(publisher)"
    }

    init(
        isbn13: String,
        isbn: String,
        title: String,
        author: String,
        publisher: String,
        coverURL: URL?,
        category: String,
        publishedAt: String,
        totalPages: Int,
        description: String
    ) {
        self.isbn13 = isbn13
        self.isbn = isbn
        self.title = title
        self.author = author
        self.publisher = publisher
        self.coverURL = coverURL
        self.category = category
        self.publishedAt = publishedAt
        self.totalPages = max(0, totalPages)
        self.description = description
    }
}

struct LibraryBook: Codable, Hashable, Identifiable, Sendable {
    let book: Book
    var status: ReadingStatus
    var recordedAt: Date

    var id: String { book.id }
}

struct BookSearchClient: Sendable {
    enum SearchError: LocalizedError, Sendable {
        case invalidResponse

        var errorDescription: String? {
            switch self {
            case .invalidResponse:
                "검색 결과를 읽지 못했어요."
            }
        }
    }

    func search(_ query: String) async throws -> [Book] {
        var components = URLComponents(string: "https://api.chaekchaek.com/api/v1/books")
        components?.queryItems = [
            URLQueryItem(name: "query", value: query),
            URLQueryItem(name: "page", value: "1")
        ]
        guard let url = components?.url else { throw SearchError.invalidResponse }

        var request = URLRequest(url: url)
        request.timeoutInterval = 10
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, 200..<300 ~= httpResponse.statusCode else {
            throw SearchError.invalidResponse
        }
        return try Self.decode(data)
    }

    static func decode(_ data: Data) throws -> [Book] {
        let response = try JSONDecoder().decode(Response.self, from: data)
        return response.items.map(\.book)
    }

    private struct Response: Decodable {
        let items: [Item]
    }

    private struct Item: Decodable {
        let title: String
        let coverImageUrl: String
        let authors: [String]
        let translators: [String]
        let publishedDate: String
        let isbn13: String
        let category: String
        let publisher: String

        var book: Book {
            Book(
                isbn13: isbn13,
                isbn: "",
                title: title,
                author: (authors + translators.map { "\($0) 옮김" }).joined(separator: " · "),
                publisher: publisher,
                coverURL: URL(string: coverImageUrl),
                category: category.split(separator: ">").last.map(String.init) ?? "",
                publishedAt: publishedDate,
                totalPages: 0,
                description: ""
            )
        }
    }
}
